결제주문 이벤트 발행 및 백업로그 적용

wanniDev opened the pull request:
지금까지는 결제 승인 결과 자체만 이벤트로 발행하고 저장하는 기능을 주로 구현하였습니다.
하지만, 결제가 실패하는 경우는 매우 다양하므로 결제에 관여하는 대부분의 과정에 이벤트를 발행하고 db에 저장하고, 만일에 대비하여 백업 로그를 저장해보는 것이 어떨까 하는 생각에서 이번 기능을 구현하고 pr을 올렸습니다.

현재 결제주문 기능의 단일 실패지점은 '카프카 브로커'로서 이 부분에 대한 'Fault tolerance'를 갖추기 위해 다음과 같은 조치를 취하였습니다.

### case1:was -------> kafka ---x---> database
was에서 발행한 이벤트 producer는 정상동작 하였으나, consumer가 제대로 동작하지 않는 경우 입니다.
본래 생성된 결제주문 이벤트는 database에 저장되어, 결제주문 페이지로 리다이렉트하는데 활용되어야 합니다. 하지만, 이 케이스로 인해, 데이터베이스에 이벤트 저장이 누락되어,  해당 페이지로 정상적으로 이동이 불가한 상태입니다.

이 경우에는 기존에 구현한 로거(#8 )에 결제주문 발행 이력을 남겼습니다.
이 발행 로그는 카프카에 문제가 생겨, 결제주문이 데이터베이스에 반영이 안되더라도, 해당 로그를 통해 제대로 진행되지 않은 결제주문에 대한 후속 조치를 시행할 수 있습니다.

### case2:was ---x---> kafka ---x---> database
카프카 시스템 자체가 예기치 못한 오류로 다운되어 이벤트를 제대로 처리하지 못하는 경우입니다.
이 경우 역시 1번 케이스와 동일합니다.

### case3:was -------> kafka -------> database(x)
마지막으로 데이터베이스가 다운되어 정상처리된 이벤트를 database에 저장을 못한 경우입니다.
이 경우는 kafka의 이벤트 로그를 활용할 수도 있고, was에 따로 백업한 로그를 활용할 수도 있겠습니다.

wanniDev committed:
Feat: paymentOrder 이벤트 등록 기능 추가

다양한 상황(장애 발생지점)에서 결제 이력을 추적하기 위해선, 결제 결과의 상태여부 뿐만 아니라, 결제를 수행하기위한 중간단계
즉, 결제주문을 생성하고 해당 결제주문의 상태 변화도 추적할 수 있어야 합니다. 그렇게해야 결제주문 상태의 올바른 변화에 따라
결제 승인요청이 처리되었는지에 대한 여부를 확인할 수 있습니다.

wanniDev committed:
Feat: PaymentOrder 이벤트 발행시 백업로그 추가

wanniDev committed:
Fix: 결제 이벤트 타입에 따라, 카프카 컨슈머에서 DomainEvent 파싱이 안되던 부분 수정

wanniDev committed:
Feat: 테스트코드로 굳이 동작시킬 필요가 없는 카프카 프로듀서 코드 부분은 test 환경이 아닐때만 동작하게 수정

wanniDev commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/tosspayments/TossPaymentsPaymentOrderProcessor.kt:
```diff
@@ -2,26 +2,45 @@ package org.collaborators.paymentslab.payment.infrastructure.tosspayments
+        paymentOrder.ready()
+        paymentOrder.pollAllEvents().forEach {
+            publisher.publishEvent(it)
+            if (!environment.activeProfiles.contains("test")) {
```
이 부분에 대해 질문이 있습니다. 이번 pr에서는 카프카의 동작 프로세스 자체는 매우 단순하고, 정말 제대로 동작하지 않는 경우는 개발자의 소스코드 문제, 혹은 시스템 문제(shutdown)가 대부분이라 판단하였습니다.

그래서, 테스트 코드가 동작하는 경우에는 카프카를 아예 배제하였는데요. 이렇게, 오픈소스 자체를 신뢰하고 테스트를 생략하는 경우가 흔한지, 아니면 어떻게든 테스트 코드를 작성하는 것이 맞는건지에 대해 여쭙고 싶습니다.

f-lab-troy replied:
위와 같이 환경에 따라 다른 비즈니스 로직을 적용하는 것은 피하는게 좋습니다.
Unit 테스트라면 kafkaTemplate을 나타내는 Mock 객체를 구현하고 주입하여 테스트하고,
Integration 테스트라면 실제로 kafka를 띄워 함께 테스트를 하는 방식으로 구현하는게 좋습니다.

wanniDev replied:
현재 f-lab의 ncp 크레딧을 신청중에 있습니다. 해당 부분은 test, dev, prod 환경의 카프카를 구성해보고 좀 더 디테일하게 진행 해보겠습니다. :)

f-lab-troy commented on app/src/main/kotlin/org/collaborators/paymentslab/support/DomainEventTranslator.kt:
```diff
@@ -16,4 +17,13 @@ class DomainEventTranslator(private val publisher: ApplicationEventPublisher) {
+    @EventListener
+    fun translate(event: PaymentOrderRecordEvent) {
+        publisher.publishEvent(
+            org.collaborators.paymentslab.log.domain.PaymentOrderRecordEvent(
```
import를 한 경우 full path reference는 안해주셔도 될 것 같습니다

wanniDev replied:
현재, 'PaymentOrderRecordEvent'는 payment 도메인에서 발행된 이벤트는 log 도메인에게 넘겨주고 있습니다. 같은 명칭의 이벤트를 하나의 클래스에 사용하다보니 'full path reference'를 불가피하게 활용하게 되었습니다.

이 부분은 바로 아래에 주신 리뷰 내용을 검토하고 나서 수정여부를 결정해보겠습니다. :)

f-lab-troy commented on log-api/log-domain/src/main/kotlin/org/collaborators/paymentslab/log/domain/PaymentOrderRecordEvent.kt:
```diff
@@ -0,0 +1,15 @@
+import org.collaborator.paymentlab.common.domain.DomainEvent
+import java.util.Date
+
+class PaymentOrderRecordEvent(
```
같은 이름의 class가 payment-domain에도 있는 이유가 있을까요?

wanniDev replied:
같은 이름의 class를 작성한 이유는 고립된 도메인간에 필요한 데이터를 넘겨주었다는 부분을 명확히 보여주기 위함이라 하면 좋을 것 같습니다.

payment-lab은 Mosuler Monolithic 아키텍처로서, 리소스 비용을 절약하면서 도메인별로 모듈을 나눠 도메인간의 고립을 유지하고 있습니다. 기존의 모놀리식의 도메인별 비즈니스 로직의 확장의 어려움과 마이크로서비스의 높은 리소스 및 유지비용을 해결하기 위해 채택한 아키텍처입니다.

현재 payment-lab에서 이벤트를 사용하는 이유들 중 가장 메인은 고립된 도메인간에 필요한 데이터를 교환하는 것입니다.
여기서, `PaymentOrderRecordEvent `를 발행하는 이유는 `PaymentOrder`가 생성된 시점을 로그로 기록하기 위함입니다. 

만약에 'PaymentOrderRecordEvent'의 네이밍을 다르게 한다면, 고립된 도메인간에 교환하는 이벤트 데이터의 대상이 무엇인지 구별하기 어려울 것 같아 이렇게 같은 명칭의 event 명을 활용하였습니다.

f-lab-troy commented on log-api/log-domain/src/main/kotlin/org/collaborators/paymentslab/log/domain/PaymentOrderRecordEvent.kt:
```diff
@@ -0,0 +1,15 @@
+    val status: String,
+    private val occurredOn: Date
+): DomainEvent {
+    override fun occurredOn(): Date {
```
occurredOn property가 override되어야 하는 이유가 궁금합니다

wanniDev replied:
DomainEvent를 구현하는 대부분의 이벤트들은 발행 시점을 활용하는 경우가 대부분일거라 판단하여 위와 같이 해당 property의 getter 역할을 하는 부분을 인터페이스의 공통함수로 지정하였습니다.

이 DomainEvent 인터페이스는 모든 도메인의 엔티티의 이벤트를 등록 및 발행 등을 수행하는 `AbstractAggregateRoot`에서 활용됩니다.
이러한 구조로 진행한 이유는 모든 엔티티의 이벤트 등록 및 발행과 관련된 로직의 중복사용을 피하기 위함입니다.

f-lab-troy commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/kafka/PaymentHistoryEventHandler.kt:
```diff
@@ -22,18 +22,27 @@ class PaymentCompleteHistoryEventHandler(
+            val eventJsonValue = record.substring(0, idx)
+            val eventType = record.substring(idx + 1)
+            val event = parseDomainEventFrom(eventJsonValue, eventType)
+            val newPaymentHistoryEntity = if (eventType == PaymentResultEvent::class.simpleName) {
```
parseDomainEventFrom과 같이 명확한 이름의 function으로 추출하는 것을 추천드립니다.

wanniDev replied:
수정하였습니다!

wanniDev committed:
Refactor: 도메인 이벤트 타입에 따라 paymentHistory를 다르게 인스턴스화하는 부분을 함수로 리팩토링

wanniDev commented:
안녕하세요. 멘토님 리뷰 검토중에 조금 고민이 있었던 부분이 생겨, 공유드립니다.
최근에 msa를 학습하며, saga pattern에 대해 학습하게 되었습니다. 해당 내용을 참고해보니 결제 승인을 포함하여, 이번 결제주문(PaymentOrder) 발행 부분의 프로세스는 **Choreography 기반의 saga pattern**과 유사하다는 걸 알게 되었습니다. 결제주문 프로세스를 도식화하자면...

![paymentlab-saga drawio](https://github.com/f-lab-edu/payment-lab/assets/81374655/8d4d958e-662d-4bdb-b1ca-c892c7a6cf5d)

위와 같이 결제주문 발행 서비스, 백업 로그 서비스는 독립적으로 동작하며, 자신의 로컬 트랜잭션을 수행하여, 결과를 이벤트로 발행하는 방식을 채택하고 있었습니다.

이러한 패턴의 특징은 '이벤트 드리븐 접근 방식', '비동기적', '도메인간의 디커플링'등이 있습니다. 해당 패턴에 대해 스터디하고, 기존 결제주문에 대해 분석을 해보니 `같은 이름의 class가 payment-domain에도 있는 이유가 있을까요?` 에 대한 멘토님의 리뷰내용이 좀더 와닿기 시작하네요.

어쩌면 도메인 서비스간의 독립적인 동작을 수행한다는 점에서 app 모듈에서 DomainEventTranslator가 별도로 필요할 이유가 없을지도 모르겠습니다.
이 부분에 대해 멘토님의 의견을 한번 더 참고해볼 수 있으면 좋을 것 같습니다. :)

f-lab-troy commented:
이전에도 대화나눈 부분이기도 하지만 제 개발 원칙의 본질은 simple(간결함)입니다. 
위의 그림에서 PaymentOrder와 PaymentOrderEvent는 결국 같은 이벤트(결제)에서 파생되는 것으로 두개를 다른 topic을 사용하는 것에 대한 정당화가 가능할지 생각해보게 됩니다. 로직의 간결성, 코드의 효율성, 유지보수, 성능 측면 등 여러면에서 토픽을 따로 나눌때 얻을 수 있는 이점이 무엇인지 설명할 수 있을지 고민해보시면 좋을 것 같습니다.

wanniDev committed:
Refactor: payment backup logger를 log-api 모듈에서 payment-api로 이동

elk 스택과 같이 logging 솔루션으로 확장하는 것을 가정하였으나, 프로젝트 특성상, 거래내역 및 과정에 대한 백업 로그를 저장하는 것외에는
별다른 기능이 요구되지 않기 때문에 비동기로 백업 로그를 챙겨서, 기존 결제 트랜잭션에 응답지연을 크게 발생시키지 않는선에서 계속 백업 로그를
출력 및 저장을 수행하면 될 것 같다.

wanniDev committed:
Remove: 불필요한 log 모듈 삭제

wanniDev committed:
Refactor: log-api 모듈에서 가지고온 payment transaction logger를 결제 승인 api에도 적용

wanniDev commented:
@f-lab-troy 안녕하세요. 멘토님, 이번 리뷰를 검토해보고 좀 더 고민해보았는데요. 이번 pr로 들어난 복잡성의 원인은 대략 3가지였던 것 같습니다.
1. 기능의 단위를 모듈의 단위로 나누고자 했던 것
2. 이벤트 드리븐에 대해 스터디하는 과정에서 EventHandler라는 워딩에 얽메여 어떻게든 `@EventListener`를 사용하려고 했던 점
3. 복잡함과 허술함의 사이의 중도를 잘 잡지 못하고 있던 점.

위 3가지 중에서 2가지를 고치고, 좀 더 단순하게 하고자, 기존의 log-api 모듈을 지우고, payment-api 모듈에 합쳤습니다. 결제 과정에 대한 백업데이터를 쌓을 뿐인데, 굳이 별도의 모듈을 구성할 필요는 없었던 것 같습니다. 

elk 스택과 같이 로깅 솔루션을 어느정도 적용하고 싶은 욕심이 있었지만, 현 시점에서는 오버엔지니어링을 하는 것에 불가할 듯 합니다.

다만, 복잡함과 허술함의 중도를 고르는 것에는 여전히 어려움을 느끼고 있습니다. 단순함에 집중하면 비즈니스 로직의 허점이 생기고 로직의 허점을 메꾸는데 집중하면 코드의 복잡성이 오릅니다...

이번 pr의 리뷰를 반영한 부분을 간단히 요약해보면, event Handling 자체를 이제는 kafka에게 위임하여, 결제에 관해서는 하나의 토픽으로 수행하는 방향으로 수정하였습니다.

추가로, 카프카와 테스트코드간의 의존성 문제와 관련된 부분은 괜찮으시다면, 별도의 pr로 작업을 수행해보겠습니다. f-lab 크레딧으로 간단하게 클라우드 서비스로 띄워볼 예정입니다 :)

wanniDev committed:
Refactor: 코드의 가독성을 높이고, 엔티티 매핑의 유연성을 확보하기 위해 결제 트랜잭션 결제 승인 요청 및 백업 로그 이벤트 컨슈머, 프로듀서 코드 리팩토링

wanniDev commented:
![paymentlab-saga drawio](https://github.com/f-lab-edu/payment-lab/assets/81374655/36e255c7-364f-47c0-b025-36aaf7b1ada8)

@f-lab-troy 안녕하세요. 멘토님, 결제주문 발급에 대해서 리뷰내용과 세션 내용을 참고하여, 주문결제 생성 api의 프로세스를 좀 더 단순화하였습니다.

이제 이벤트 핸들링은 카프카에게 온전히 위임하였고, 카프카 브로커가 단일 실패지점이라는 점을 인지하고, 결제주문 발행 시점에 백업 로그를 확보한 다음 카프카 파티션에 레코드를 produce 하는 방식을 채택하였습니다.

또한, 카프카 컨슈머를 수행하는 이벤트 핸들링 코드(PaymentCompletedEvent.kt) 부분도 리팩토링을 수행하였습니다.

여전히 if-else를 통해 DomainEvent의 구현체를 구분하는 코드는 있지만, 해당 부분은  '능동적인 객체'에게 위임함으로써 적어도 관리하는 이벤트의 종류가 증가하여도 극단적으로 조건부 브렌치가 길어지진 않을 것 같습니다.

f-lab-troy commented on payment-api/payment-domain/src/main/kotlin/org/collaborators/paymentslab/payment/domain/entity/PaymentHistory.kt:
```diff
@@ -68,6 +83,15 @@ class PaymentHistory protected constructor(
         }
+
+        fun newInstanceFrom(domainEvent: DomainEvent): PaymentHistory {
+            if (domainEvent is PaymentOrderRecordEvent) {
```
If else가 없이 newInstanceFrom(domainEvent)만 call하면 안되는 이유가 있을까요?

wanniDev replied:
안녕하세요! 멘토님, newInstanceFrom(domainEvent)를 그대로 콜하게 될경우, 함수 네이밍과 함수 매개변수의 타입이 일치하여, 결과적으로 별도의 baseCase없이 재귀함수를 무한하게 호출하는 형태가 되어, stackOverflow가 발생하게 되었습니다.

그래서 그런 오류가 발생하는 상황을 피하면서, 기존의 코드보다 조금이라도 간결하게 개선해보고자 코틀린의 스마트캐스트 기능을 활용하였습니다.

wanniDev committed:
Reafactor: 가독성을 높이기위해 if-else 를 when 구문으로 변경
